pkg/operator/options/options.go

```diff
@@ -0,0 +1,177 @@
+	o.setFlags = map[string]bool{}
+	cliFlags := sets.New[string]()
+	fs.Visit(func(f *flag.Flag) {
+		cliFlags.Insert(f.Name)
+	})
+	fs.VisitAll(func(f *flag.Flag) {
+		envName := strings.ReplaceAll(strings.ToUpper(f.Name), "-", "_")
+		_, ok := os.LookupEnv(envName)
+		o.setFlags[f.Name] = ok || cliFlags.Has(f.Name)
+	})
+
+	if err := o.Validate(); err != nil {
+		return fmt.Errorf("validating options, %w", err)
+	}
+
+	// if ClusterID is not set, generate it from cluster endpoint
+	// if clusterEndpoint is empty, it might awaiting merge from MergeSettings(), do not run GetAPIServerName() or it could panic
+	// TODO: chore: remove clusterEndpoint validation logic here when karpenter-global-settings (and merge logic) are completely removed
+	if o.ClusterID == "" && o.ClusterEndpoint != "" {
+		o.ClusterID = getAKSClusterID(o.GetAPIServerName())
+	}
+
+	return nil
+}
+
+func (o *Options) ToContext(ctx context.Context) context.Context {
+	return ToContext(ctx, o)
+}
+
+func (o *Options) MergeSettings(ctx context.Context) {
+	s := settings.FromContext(ctx)
+	mergeField(&o.ClusterName, s.ClusterName, o.setFlags["cluster-name"])
+	mergeField(&o.ClusterEndpoint, s.ClusterEndpoint, o.setFlags["cluster-endpoint"])
+	mergeField(&o.VMMemoryOverheadPercent, s.VMMemoryOverheadPercent, o.setFlags["vm-memory-overhead-percent"])
+	mergeField(&o.ClusterID, s.ClusterID, o.setFlags["cluster-id"])
+	mergeField(&o.KubeletClientTLSBootstrapToken, s.KubeletClientTLSBootstrapToken, o.setFlags["kubelet-client-tls-bootstrap-token"])
+	mergeField(&o.SSHPublicKey, s.SSHPublicKey, o.setFlags["ssh-public-key"])
+	mergeField(&o.NetworkPlugin, s.NetworkPlugin, o.setFlags["network-plugin"])
+	mergeField(&o.NetworkPolicy, s.NetworkPolicy, o.setFlags["network-policy"])
+	mergeField(&o.NodeIdentities, s.NodeIdentities, o.setFlags["node-identities"])
+
+	if err := o.validateRequiredFields(); err != nil {
+		panic(fmt.Errorf("checking required fields, %w", err))
+	}
+
+	// if ClusterID is not set, generate it from cluster endpoint
+	if o.ClusterID == "" {
+		o.ClusterID = getAKSClusterID(o.GetAPIServerName())
+	}
+}
+
+func ToContext(ctx context.Context, opts *Options) context.Context {
+	return context.WithValue(ctx, optionsKey{}, opts)
+}
+
+func FromContext(ctx context.Context) *Options {
+	retval := ctx.Value(optionsKey{})
+	if retval == nil {
+		return nil
+	}
+	return retval.(*Options)
+}
+
+func mergeField[T any](dest *T, src T, isDestSet bool) {
+	if !isDestSet {
+		*dest = src
+	}
+}
+
+// getAKSClusterID returns cluster ID based on the DNS prefix of the cluster.
+// The logic comes from AgentBaker and other places, originally from aks-engine
+// with the additional assumption of DNS prefix being the first 33 chars of FQDN
+func getAKSClusterID(apiServerFQDN string) string {
+	dnsPrefix := apiServerFQDN[:33]
+	h := fnv.New64a()
+	h.Write([]byte(dnsPrefix))
+	r := rand.New(rand.NewSource(int64(h.Sum64()))) //nolint:gosec
+	return fmt.Sprintf("%08d", r.Uint32())[:8]
+}
```

Comment:
This seems copy pasted from settings, we have it in both places and in both places it isn't tested anywhere. Not that its easy to test the hashing but we can probably just use it from one place

Reply:
The one in `Settings` will be removed once we drop the support, with this one only remains.